### PR TITLE
Add tests for NodeStack and rework

### PR DESCRIPTION
**Pull request by lorduke22**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/datastructures/stacks/NodeStack.java [hunk 1]
```diff
@@ -3,159 +3,65 @@
...
+        return size;
     }
 
-    /**
-     * Print the contents of the stack in the following format.
-     *
-     * <p>
-     * x <- head (next out) y z <- tail (first in) . . .
-     */
     public void print() {
```
Comment (siriak) [inline src/main/java/com/thealgorithms/datastructures/stacks/NodeStack.java:59]: Please remove print function, we don't provide visualization means, only the data structure itself

**Review by siriak — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/datastructures/stacks/NodeStack.java (hunk 1 above)
Comment (lorduke22) [reply src/main/java/com/thealgorithms/datastructures/stacks/NodeStack.java:59]: ok no problem...sorry

**Review by siriak — APPROVED** [review]
Looks good, thanks!

**Review by siriak — CHANGES_REQUESTED** [review]
Please resolve conflicts

File: src/main/java/com/thealgorithms/datastructures/stacks/NodeStack.java (hunk 1 above)
Comment (siriak) [reply src/main/java/com/thealgorithms/datastructures/stacks/NodeStack.java:102]: It's still here

**Review by siriak — CHANGES_REQUESTED** [review]
Everything except print looks good

**Review by siriak — CHANGES_REQUESTED** [review]
Please fix PR check

**Comment by lorduke22** [comment]
I cannot solve this PR because during a CI/CD pipeline run with Infer, a potential Null Dereference was detected in the BinaryTreeTest class. 
The analysis flagged the following line as a potential issue:

**src/test/java/com/thealgorithms/datastructures/trees/BinaryTreeTest.java:38: error: Null Dereference(NULLPTR_DEREFERENCE)
  `BinaryTree$Node BinaryTree.getRoot()` could be null (from the call to `BinaryTree()` on line 28) and is dereferenced.**

To solve the build problem, if I wish, I can apply a temporary fix to prevent the test from failing on the BinaryTreeTest class (e.g., by adding a null check in getRoot())

@siriak Let me know how to proceed

**Comment by siriak** [comment]
Hmm, that's odd. Could you try to fix the issue here or in a separate PR?

**Comment by Chiefpatwal** [comment]
> I cannot solve this PR because during a CI/CD pipeline run with Infer, a potential Null Dereference was detected in the BinaryTreeTest class. The analysis flagged the following line as a potential issue:
> 
> **src/test/java/com/thealgorithms/datastructures/trees/BinaryTreeTest.java:38: error: Null Dereference(NULLPTR_DEREFERENCE) `BinaryTree$Node BinaryTree.getRoot()` could be null (from the call to `BinaryTree()` on line 28) and is dereferenced.**
> 
> To solve the build problem, if I wish, I can apply a temporary fix to prevent the test from failing on the BinaryTreeTest class (e.g., by adding a null check in getRoot())
> 
> @siriak Let me know how to proceed

Hey @lorduke22, I’m facing the same conflict issue due to the failing Infer run check. I’m stuck and could use some guidance if you have any insights on resolving it.

**Comment by lorduke22** [comment]
@siriak I should have solved the problem in this PR.
Hi @Chiefpatwal, the problem lies in the BinaryTreeTest class I see the getRoot() method may return null if the root node has been removed and there are no more nodes in the tree. This is why Infer reports a potential NullPointerException when attempting to access the date field in t.getRoot().data.
To resolve the error, I added a null check on getRoot() in test2.

image

Feel free to add the following control to your code

**Review by siriak — APPROVED** [review]
Great, thanks for fixing this!

**Pull request merged**